diff --git a/internal/race/norace.go b/internal/race/norace.go
index fc6e13f..346564b 100644
--- a/internal/race/norace.go
+++ b/internal/race/norace.go
@@ -8,8 +8,6 @@ package race

 import (
 	"unsafe"
-
-	"github.com/CodSpeedHQ/codspeed-go/testing/internal/abi"
 )

 const Enabled = false
@@ -35,8 +33,8 @@ func Read(addr unsafe.Pointer) {
 func ReadPC(addr unsafe.Pointer, callerpc, pc uintptr) {
 }

-func ReadObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr) {
-}
+// func ReadObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr) {
+// }

 func Write(addr unsafe.Pointer) {
 }
@@ -44,8 +42,8 @@ func Write(addr unsafe.Pointer) {
 func WritePC(addr unsafe.Pointer, callerpc, pc uintptr) {
 }

-func WriteObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr) {
-}
+// func WriteObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr) {
+// }

 func ReadRange(addr unsafe.Pointer, len int) {
 }
diff --git a/internal/race/race.go b/internal/race/race.go
index d345009..ef54ea5 100644
--- a/internal/race/race.go
+++ b/internal/race/race.go
@@ -7,7 +7,6 @@
 package race

 import (
-	"github.com/CodSpeedHQ/codspeed-go/testing/internal/abi"
 	"unsafe"
 )

@@ -36,8 +35,8 @@ func Read(addr unsafe.Pointer)
 //go:linkname ReadPC
 func ReadPC(addr unsafe.Pointer, callerpc, pc uintptr)

-//go:linkname ReadObjectPC
-func ReadObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr)
+// //go:linkname ReadObjectPC
+// func ReadObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr)

 //go:linkname Write
 func Write(addr unsafe.Pointer)
@@ -45,8 +44,8 @@ func Write(addr unsafe.Pointer)
 //go:linkname WritePC
 func WritePC(addr unsafe.Pointer, callerpc, pc uintptr)

-//go:linkname WriteObjectPC
-func WriteObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr)
+// //go:linkname WriteObjectPC
+// func WriteObjectPC(t *abi.Type, addr unsafe.Pointer, callerpc, pc uintptr)

 //go:linkname ReadRange
 func ReadRange(addr unsafe.Pointer, len int)
